Migration to pyproject.toml and tox

JiriPavela opened the pull request:
This PR removes the obsolete setup.py project configuration and substitutes it with the new pyproject.toml configuration file which is specified by multiple PEPs. The pyproject.toml also contains tool-specific configuration (e.g., for pylint, coverage, ...).

Moreover, a tox configuration has been created and tested so that we can easily automate testing, linting and type checking both locally and in GH Actions. Currently, the only python version supported by tox is 3.8 due to complicated dependency conflicts in higher python versions.

This PR solves #92 and partially #95 (GH Actions are not updated to use Tox yet).

JiriPavela committed:
Add the first version of pyproject.toml

This commit adds a new pyproject.toml file for packaging Perun and
removes the old setup.py file.

JiriPavela committed:
Hotfix memory collector issue breaking tests.

For some reason, closing the log file in the finalize function
segfaults on _int_malloc. Let the OS do the file closing then.

JiriPavela committed:
Fix AAT degradation method warning.

The method was relying on a deprecated default value for pandas
.mean() method. This fix explicitly sets the numeric_only parameter.

JiriPavela committed:
Fix fuzzing timeout and coverage tests.

The timeout test was using a bit too high timeout. As a consequence,
the timeout was not reached on fast machines.

The coverage test has stopped working for GCC 11+ which changed the
naming conventions for dump files.

JiriPavela committed:
Fix test profiles not containing b2 coefficient.

Some of the example profiles used in tests were modified in the
past to reflect changes made in quadratic regression analysis.
However, not all relevant profiles were modified. This commit
adds the b2 coefficient to model entries.

JiriPavela committed:
Fix reference systemtap scripts.

A sampling hotfix changed the resulting systemtap scripts and the
corresponding tests were not updated.

JiriPavela committed:
Update pyproject.toml and add tox configuration.

Tox is currently configured such that Perun is tested only for
Python3.8. For higher python versions, the custom wheels have to
be either updated or removed as some dependencies are in conflict.

JiriPavela committed:
Extend the tox conf with additional envs.

The tox now automates linter checks, static type checking and
generating documentation.

The optional dependencies were reworked to avoid duplicity when
specifying 'deps' in tox.

tfiedor commented on perun/collect/memory/malloc.c:
```diff
@@ -168,7 +168,8 @@ __attribute__ ((constructor)) void initialize (void) {
     if(logFile != NULL){
         fprintf(logFile, "EXIT %fs\n", clock() / (double)CLOCKS_PER_SEC);
-        fclose(logFile);
+        // This is causing segfaults for some reason, hotfix
```
```suggestion
        // FIXME: This is causing segfaults for some reason, hotfix
```
This way it won't go fully ingored.

JiriPavela replied:
Fixed.

JiriPavela replied:
I have updated the README instructions a bit. Most of the instructions remained as they were, as the installation is usually done through the Makefile, which was updated accordingly. However, I have removed the outdated `perun init` command in both install/dev, and included some basic Tox instructions.

tfiedor commented on pyproject.toml:
```diff
@@ -0,0 +1,112 @@
+[project]
+name = "perun"
+description = "Perun: Lightweight Performance Version System"
+version = "0.20.4"
```
Is it possible to extract the version somehow? It's is not ideal for it to be as hardcoded.

JiriPavela replied:
Good catch, seems like I forgot to de-duplicate the versions before doing the PR. So the `pyproject.toml` file is actually the preferred place to hard-code the package version, as far as I understand. Project/package metadata should be stored in the toml file to make them available for package distribution software. So instead, I have made the `__init__.py` `__version__` attribute dynamic such that it reads the version from the installed package (see [this SO answer](https://stackoverflow.com/a/72168209)).

tfiedor reviewed (approved):
I think this is fine, thank you for solving this issue.

However, I would ask you one more change: please change README according to these changes, since now building, running of tests, etc. has maybe changed.

JiriPavela committed:
Add a FIXME keyword to the malloc.c issue.

JiriPavela committed:
Link __init__ and package versions.

Project version should be defined in pyproject.toml as part of the
package metadata. However, our __init__ module was hardcoding the
version, resulting in duplicate version specification. This commit
fixes this.

JiriPavela committed:
Update README install/development instructions.

tfiedor commented on README.md:
```diff
@@ -191,6 +189,18 @@ by the installation.
+
+    tox run -e pyXY
+
+where `XY` is the python version you wish to test. 
```
Suggestion: write this as `py3.X` maybe? and mention which version we currently safely support?

JiriPavela replied:
The `pyXY` is actually a naming convention used by Tox to automatically deduce the Python interpreter version (see https://tox.wiki/en/latest/config.html#basepython), but it seems that it understands `pyX.Y` as well (no mention of this in docs tho?), so I updated it. As for the supported versions, I have updated the README with the command to show available Tox environments.

tfiedor replied:
OK, that's fine then.

JiriPavela committed:
Update Tox environment name for tests.
